nuxeo-compound-documents/src/test/java/org/nuxeo/compound/documents/TestCompoundDocumentRest.java

```diff
@@ -43,57 +51,84 @@
     public void testCreateCompoundDocumentFolder() throws IOException {
-        session.createDocument(session.createDocumentModel("/", "myCompoundDocument", "CompoundDocument"));
+        session.createDocument(session.createDocumentModel("/", "myCompoundDocument", COMPOUND_DOCTYPE));
         txFeature.nextTransaction();
-        String data = "{\"entity-type\": \"document\",\"type\": \"CompoundDocumentFolder\", \"name\": \"myCompoundDocumentFolder\"}";
+        String data = "{\"entity-type\": \"document\",\"type\": \"" + COMPOUND_FOLDER_DOCTYPE
+                + "\", \"name\": \"myCompoundDocumentFolder\"}";
         try (CloseableClientResponse response = getResponse(RequestType.POST, "path/myCompoundDocument", data)) {
             assertEquals(SC_CREATED, response.getStatus());
         }
         var headers = Map.of(MarshallingConstants.EMBED_ENRICHERS + ".document", SubtypesJsonEnricher.NAME);
         try (CloseableClientResponse response = getResponse(RequestType.GET,
                 "path/myCompoundDocument/myCompoundDocumentFolder", headers)) {
-            assertCompoundResponse(response, "Folderish");
+            assertCompoundResponse(response, Set.of("Folderish"));
         }
     }
 
-    protected void assertCompoundResponse(CloseableClientResponse response, String expectedFacet) throws IOException {
+    protected void assertCompoundResponse(CloseableClientResponse response, Set<String> expectedFacets)
+            throws IOException {
         assertEquals(SC_OK, response.getStatus());
         JsonNode node = mapper.readTree(response.getEntityInputStream());
-        assertEquals(1, node.get("facets").size());
-        assertEquals(expectedFacet, node.get("facets").get(0).asText());
+        JsonNode facets = node.get("facets");
+        assertEquals(expectedFacets.size(), facets.size());
+        facets.forEach(f -> assertTrue(expectedFacets.contains(f.asText())));
         var actualSubtypes = node.get("contextParameters").get("subtypes");
         assertEquals(5, actualSubtypes.size());
-        var allowedSubtypes = Set.of("File", "Picture", "Video", "Audio", "CompoundDocumentFolder");
+        var allowedSubtypes = Set.of("File", "Picture", "Video", "Audio", COMPOUND_FOLDER_DOCTYPE);
         actualSubtypes.forEach(t -> assertTrue(allowedSubtypes.contains(t.get("type").asText())));
     }
+
+    @Test
+    public void testCompoundDocumentUpload() throws IOException {
+        String batchId;
+        try (CloseableClientResponse response = getResponse(RequestType.POST, "upload")) {
+            JsonNode node = mapper.readTree(response.getEntityInputStream());
+            batchId = node.get("batchId").asText();
+        }
+        Builder builder = service.path("upload/" + batchId + "/0")
+                                 .accept(MediaType.APPLICATION_JSON)
+                                 .header("X-File-Name", "test");
+        try (FileInputStream in = new FileInputStream(makeTestZipBlob().getFile());
```

Comment:
You should declare the in variable as `InputStream` because it is the interface, it is the same thing than List/ArrayList or Set/HashSet.
Or, on a java 11 way, just use var here as we don't really care 😄.